notebooks/norse_tutorial.ipynb

```diff
@@ -1,868 +1,1010 @@
+    "        loss.backward()\n",
+    "        optimizer.step()\n",
+    "        losses.append(loss.item())\n",
+    "\n",
+    "    mean_loss = np.mean(losses)\n",
+    "    return losses, mean_loss"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "KtdcQi_18Xip"
+   },
+   "source": [
+    "Just like the training function, the test function is standard boilerplate, common with any other supervised learning task."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {
+    "colab": {},
+    "colab_type": "code",
+    "id": "Gca4ZzatApWD"
+   },
+   "outputs": [],
+   "source": [
+    "def test(model, device, test_loader, epoch):\n",
+    "    model.eval()\n",
+    "    test_loss = 0\n",
+    "    correct = 0\n",
+    "    with torch.no_grad():\n",
+    "        for data, target in test_loader:\n",
+    "            data, target = data.to(device), target.to(device)\n",
+    "            output = model(data)\n",
+    "            test_loss += torch.nn.functional.nll_loss(\n",
+    "                output, target, reduction=\"sum\"\n",
+    "            ).item()  # sum up batch loss\n",
+    "            pred = output.argmax(\n",
+    "                dim=1, keepdim=True\n",
+    "            )  # get the index of the max log-probability\n",
+    "            correct += pred.eq(target.view_as(pred)).sum().item()\n",
+    "\n",
+    "    test_loss /= len(test_loader.dataset)\n",
+    "\n",
+    "    accuracy = 100.0 * correct / len(test_loader.dataset)\n",
+    "\n",
+    "    return test_loss, accuracy"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "r4sGruNn8w49"
+   },
+   "source": [
+    "As is the the outer training loop."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {
+    "colab": {
+     "base_uri": "https://localhost:8080/",
+     "height": 392
     },
+    "colab_type": "code",
+    "id": "wU-b7Q8eBVca",
+    "outputId": "7469a1a5-7ac9-4675-8c20-c91c04c1a1f0"
+   },
+   "outputs": [
     {
-      "cell_type": "markdown",
-      "metadata": {
-        "id": "MOYVGOvhrDty",
-        "colab_type": "text"
-      },
-      "source": [
-        "## MNIST dataset"
-      ]
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "100%|██████████| 1875/1875 [07:43<00:00,  4.05it/s]\n"
+     ]
     },
     {
-      "cell_type": "code",
-      "metadata": {
-        "id": "RPs81D-QrFWV",
-        "colab_type": "code",
-        "outputId": "44a63b0a-8aec-4eb0-b954-649e609d5d09",
-        "colab": {
-          "base_uri": "https://localhost:8080/",
-          "height": 255
-        }
-      },
-      "source": [
-        "import torchvision\n",
-        "\n",
-        "BATCH_SIZE = 32\n",
-        "\n",
-        "transform = torchvision.transforms.Compose(\n",
-        "    [\n",
-        "        torchvision.transforms.ToTensor(),\
```

Comment:
While it is definitely good that you ran the notebook to conclusion. The fact that we only reach 98% is actually a bug, because in internal code I easily reached >99%. I'm looking into a fix for that now.

Reply:
Interesting! I didn't even look into the code.
I added a brief summary section in the end to conclude/summarise/perspectivise. Any new results that are competing with state-of-the-art would be great to include there!

Reply:
I agree, that we can address this in a future patch. I've now played a bit with the network architecture, maybe that was the cause of the discrepancy. I will upload a new version of the tutorial notebook, once I've figured it out.